src/cart/src/cart/crt_ctl.c

```diff
@@ -121,15 +118,12 @@ crt_ctl_get_uri_cache_size_cb(d_list_t *rlink, void *arg)
 	D_ASSERT(arg != NULL);
 
 	ui = crt_ui_link2ptr(rlink);
-
-	D_MUTEX_LOCK(&ui->ui_mutex);
 	for (i = 0; i < CRT_SRV_CONTEXT_NUM; i++) {
-		if (ui->ui_uri[i] == NULL)
+		if (atomic_load_relaxed(&ui->ui_uri[i]) == NULL)
```

Comment:
This change doesn't work at all actually, if you look at what the wider function is doing it needs to hold the lock the whole time, it calls traverse to find the size, then again to populate the array.  Using atomics like this the size might change between the two calls, so the only way to do this would be to traverse once and do a realloc per cb, not ideal but this isn't performance critical code.  Either that or you could set a size, and if it grows then just return a partial sample (which is OK anyway, as any sample will be potentally out-of-date once the lock was dropped).

Reply:
Agree.